src/dataflow/src/source/mod.rs

```diff
@@ -1766,6 +1795,8 @@ impl SourceReaderPersistence {
                 .or_default() += diff;
```

Comment:
I know this is not the current change, but I'm struggling to convince myself that always compacting all of the bindings up like this is correct because my understanding is that this compaction is as aggressive as possible, and will not respect the since frontier, which could be held back if there was an exactly once sink? its also possible that I'm missing something here

Reply:
It's complicated. :sweat_smile: But, as always, it's good that you ask, because it forces me to re-think, and condense my thoughts such that someone else now also knows how this works.

The core reason is that compacting a bunch of differential updates (aka advancing the since) and compacting the timestamp bindings (in the history) are different things.

I'll try and explain using an example. For different steps, we will both look at the current contents of the timestamp histories and how that is reflected in differential updates that can be used to restore that state.

(`tx` is `timestamp x`, `ox` is `offset x`, `ctx` is `capability time x`, note how the time at which updates happen is decoupled from the time that timestamp bindings refer to)

```
State 0:
ts_histories: [(t10, o5), (t20, o10)]
all emitted differential updates:
((t10, o5), ct0, 1)
((t20, o10), ct0, 1)
```

let's allow compaction up to `t30`.
This fill first advance all times, yielding `ts_histories: [(t30, o5), (t30, o10)]`, and then they will be combined to `ts_histories: [(t30, o10)]`
```
State 1:
ts_histories: [(t30, o10)]
all emitted differential updates:
((t10, o5), ct0, 1)
((t20, o10), ct0, 1)
((t10, o5), ct1, -1)
((t20, o10), ct1, -1)
((t30, o10), ct1, 1)
```

We have to full fidelity of changes to the timestamp histories in the differential updates. If we want to restore the state up to time `c1`, we would have to "sum up" all the updates that are `<=c1`, ignoring their timestamp. Same goes for restoring at `c0`. And we don't care about the compaction frontier (aka since) here, because that is implicit in the timestamp bindings that are still present.

Side note: allowing compaction (i.e. advancing the since) on the collection that stores the differential binding updates means we now loose historical resolution. If we allowed compaction up to `ct1`, the `-1`/`1` would anihilate. But it's not the same thing as "compacting" the timestamp histories.

Maybe that helps, but I'm very happy to do another call.